monitoring/storage.go

```diff
@@ -0,0 +1,238 @@
+		}
+
+		parent := filepath.Dir(p)
+		if parent == p {
+			// shouldn't happen: root reached and it's not a dir
+			return trace.BadParameter("%s is root and is not a dir", p)
+		}
+		p = parent
+	}
+}
+
+func (c *StorageChecker) checkFsType(ctx context.Context, reporter health.Reporter) error {
+	if len(c.Filesystems) == 0 {
+		return nil
+	}
+
+	mnt, err := fsFromPath(c.path)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	probe := &pb.Probe{Checker: c.Name()}
+
+	if utils.StringInSlice(c.Filesystems, mnt.SysTypeName) {
+		probe.Status = pb.Probe_Running
+	} else {
+		probe.Status = pb.Probe_Failed
+		probe.Detail = fmt.Sprintf("path %s requires %v, belongs to %s mount point of type %s",
+			c.Path, c.Filesystems, mnt.DirName, mnt.SysTypeName)
+	}
+	reporter.Add(probe)
+	return nil
+}
+
+type childPathFirst []sigar.FileSystem
+
+func (a childPathFirst) Len() int           { return len(a) }
+func (a childPathFirst) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a childPathFirst) Less(i, j int) bool { return strings.HasPrefix(a[i].DirName, a[j].DirName) }
+
+func fsFromPath(path string) (*sigar.FileSystem, error) {
+	cleanpath, err := filepath.EvalSymlinks(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	mounts := sigar.FileSystemList{}
+	err = mounts.Get()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sort.Sort(childPathFirst(mounts.List))
+
+	for _, mnt := range mounts.List {
+		if strings.HasPrefix(cleanpath, mnt.DirName) {
+			return &mnt, nil
+		}
+	}
+
+	return nil, trace.BadParameter("failed to locate filesystem for %s", path)
+}
+
+func (c *StorageChecker) checkCapacity(ctx context.Context, reporter health.Reporter) error {
+	var stat syscall.Statfs_t
+
+	err := syscall.Statfs(c.path, &stat)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	avail := uint64(stat.Bsize) * stat.Bavail
+	if avail < c.MinFreeBytes {
+		reporter.Add(&pb.Probe{
+			Checker: c.Name(),
+			Detail: fmt.Sprintf("%s available space left on %s, minimum of %s required",
+				humanize.Bytes(avail), c.Path, humanize.Bytes(c.MinFreeBytes)),
+			Status: pb.Probe_Failed,
+		})
+	}
+
+	return nil
+}
+
+func (c *StorageChecker) checkWriteSpeed(ctx context.Context, reporter health.Reporter) (err error) {
+	if c.MinBytesPerSecond == 0 {
+		return
+	}
+
+	var file *os.File
+	file, err = ioutil.TempFile(c.path, "probe")
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	defer func() {
+		err_c := file.Close()
+		err_r := os.Remove(file.Name())
+		err = trace.NewAggregate(err_c, err_r)
+	}()
+
+	start := time.Now()
+	for i := 0; i < cycles; i++ {
+		buf := make([]byte, blockSize)
+		_, err = file.Write(buf)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if ctx.Err() != nil {
+			return trace.Wrap(ctx.Err())
+		}
+	}
+	err = file.Sync()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	elapsed := time.Since(start).Seconds()
+	bps := uint64(blockSize * cycles / elapsed)
+
+	if bps >= c.MinBytesPerSecond {
+		return nil
```

Comment:
Don't we need to add a successful probe in this case to indicate the check passed? I think we do that in other places?

Reply:
it's there, see line 75